### PR TITLE
Show and edit date in the same timezone

### PR DESCRIPTION
**Pull request by peter-gribanov**
## Subject

The `date` and `datetime` field has a `timezone` option that is used as second argument for Twig's `date` function ([doc](https://symfony.com/doc/master/bundles/SonataAdminBundle/reference/action_list.html#available-types-and-associated-options)).

It seems wrong to me that the date is displayed in one time zone, and is being live-edited in another time zone. Due to the difference between the server time zone and the time zone specified in the options, we can get completely different dates, which can lead to administrative errors.

I assume that the time zone is used so that the administrator can specify his time zone and not look at all the dates in the server time zone. This is useful because JavaScript calendars do not know in which time zone the server is located and display the calendar in the user's time zone. If the time zone of the server and the user are different, this can also lead to problems.

## Changelog

```markdown
### Fixed
- Show and edit `date` and `datetime` in the same timezone
```

### CODE REVIEW

File: src/Action/SetObjectFieldValueAction.php [hunk 1]
```diff
@@ -115,7 +115,20 @@ public function __invoke(Request $request): JsonResponse
 
         // Handle date and datetime types have setter expecting a DateTime object
         if ('' !== $value && \in_array($fieldDescription->getType(), ['date', 'datetime'], true)) {
-            $value = new \DateTime($value);
+            $inputTimezone = new \DateTimeZone(date_default_timezone_get());
+            $outputTimezone = $fieldDescription->getOption('timezone');
+
+            if (\is_string($outputTimezone)) {
```
Comment (phansys) [inline src/Action/SetObjectFieldValueAction.php:121]: IMO, the validation about allowed types should be done at field description.

File: src/Action/SetObjectFieldValueAction.php [hunk 2]
```diff
@@ -115,7 +115,20 @@ public function __invoke(Request $request): JsonResponse
 
         // Handle date and datetime types have setter expecting a DateTime object
         if ('' !== $value && \in_array($fieldDescription->getType(), ['date', 'datetime'], true)) {
-            $value = new \DateTime($value);
+            $inputTimezone = new \DateTimeZone(date_default_timezone_get());
```
Comment (core23) [inline src/Action/SetObjectFieldValueAction.php:118]: Please write a test that will test a field timezone.

Comment (phansys) [inline src/Action/SetObjectFieldValueAction.php:118]: For the current underlying timezone, I think we could simply use `null`.
[`DateTime::__construct()`](https://www.php.net/manual/en/datetime.construct.php#refsect1-datetime.construct-parameters):
>timezone
>A DateTimeZone object representing the timezone of $time.
>
>If $timezone is omitted, the current timezone will be used.

File: src/Action/SetObjectFieldValueAction.php (hunk 1 above)
Comment (peter-gribanov) [reply src/Action/SetObjectFieldValueAction.php]: Allowed values for timezone option is `\DateTimeZone|string|false|null`
https://github.com/twigphp/Twig/blob/3.x/src/Extension/CoreExtension.php#L410

File: src/Action/SetObjectFieldValueAction.php (hunk 2 above)
Comment (peter-gribanov) [reply src/Action/SetObjectFieldValueAction.php:118]: But we do not need a date in the current time zone. We need to convert the date in the user time zone to the system (default) time zone.

Comment (phansys) [reply src/Action/SetObjectFieldValueAction.php:118]: From the documentation perspective, "current timezone" is actually what `date_default_timezone_get()` returns.

File: src/Action/SetObjectFieldValueAction.php [hunk 3]
```diff
@@ -115,7 +115,18 @@ public function __invoke(Request $request): JsonResponse
...
+
+            if (!$outputTimezone instanceof \DateTimeZone) {
+                $outputTimezone = new \DateTimeZone($outputTimezone);
+            }
+
+            $value = new \DateTime($value, $outputTimezone);
+
+            if ($outputTimezone->getName() !== $inputTimezone->getName()) {
+                $value->setTimezone($inputTimezone);
+            }
```
Comment (phansys) [inline src/Action/SetObjectFieldValueAction.php:129]:
```suggestion
            if ($timezone = $fieldDescription->getOption('timezone')) {
                $inputTimezone = $timezone instanceof \DateTimeZone ? $timezone : new \DateTimeZone($timezone);
                $value->setTimezone($inputTimezone);
            }
```

File: tests/Action/SetObjectFieldValueActionTest.php [hunk 4]
```diff
@@ -126,7 +126,22 @@ public function testSetObjectFieldValueAction(): void
         $this->assertSame(Response::HTTP_OK, $response->getStatusCode());
     }
 
-    public function testSetObjectFieldValueActionWithDate(): void
+    public function getTimeZones(): array
```
Comment (phansys) [inline tests/Action/SetObjectFieldValueActionTest.php:129]:
```suggestion
    public function getTimeZones(): iterable
```

File: src/Action/SetObjectFieldValueAction.php (hunk 3 above)
Comment (peter-gribanov) [reply src/Action/SetObjectFieldValueAction.php:129]: This is the wrong code.

Comment (phansys) [reply src/Action/SetObjectFieldValueAction.php:129]: Maybe I'm missing something. My suggestion is based on this fact: https://3v4l.org/UF1Rp.
How calling `date_default_timezone_get()` is different?

Comment (peter-gribanov) [reply src/Action/SetObjectFieldValueAction.php]: Your code does not do what is needed here.

Comment (peter-gribanov) [reply src/Action/SetObjectFieldValueAction.php]: Please review again.

Comment (peter-gribanov) [reply src/Action/SetObjectFieldValueAction.php]: Besides. Assigning a value to a variable in a condition is a sign of a bad coding style. It is worth avoiding this.

Comment (phansys) [reply src/Action/SetObjectFieldValueAction.php]: There is no need to perform the assign operation inside the `if` condition :+1:. Nevertheless, that isn't changing the result.
I'll be back with some feedback when I'm able to check it locally.

Comment (phansys) [reply src/Action/SetObjectFieldValueAction.php]: I've executed the test `SetObjectFieldValueActionTest` locally, and I can't find any difference in the results using the original code or your changes.
I'd suggest to create a representative test exposing what this PR fixes.

**Comment by phansys** [comment]
Hi @peter-gribanov!
Did you noticed the comment I left a week ago? (https://github.com/sonata-project/SonataAdminBundle/pull/6149#discussion_r441970936).

**Comment by peter-gribanov** [comment]
> Did you noticed the comment I left a week ago?

I missed it. I'm sorry.

File: src/Action/SetObjectFieldValueAction.php (hunk 3 above)
Comment (peter-gribanov) [reply src/Action/SetObjectFieldValueAction.php]: I have added tests for this change.

**Comment by peter-gribanov** [comment]
@phansys i have added tests for this change.

**Comment by VincentLanglet** [comment]
@peter-gribanov You should rebase your branch since we added Phpstan.

File: tests/Action/SetObjectFieldValueActionTest.php [hunk 5]
```diff
@@ -171,9 +190,20 @@ public function testSetObjectFieldValueActionWithDate(): void
...
 
         $this->assertSame(Response::HTTP_OK, $response->getStatusCode());
+
+        $default_timezone = new \DateTimeZone(date_default_timezone_get());
+        $expected_date = new \DateTime($request->query->get('value'), $expected_timezone);
+        $expected_date->setTimezone($default_timezone);
+
+        $this->assertInstanceOf(\DateTime::class, $object->getDateProp());
+        $this->assertSame($expected_date->format('Y-m-d'), $object->getDateProp()->format('Y-m-d'));
+        $this->assertSame($default_timezone->getName(), $object->getDateProp()->getTimezone()->getName());
```
Comment (phansys) [inline tests/Action/SetObjectFieldValueActionTest.php:200]:
```suggestion
        $defaultTimezone = new \DateTimeZone(date_default_timezone_get());
        $expectedDate = new \DateTime($request->query->get('value'), $expectedTimezone);
        $expectedDate->setTimezone($defaultTimezone);

        $this->assertInstanceOf(\DateTime::class, $object->getDateProp());
        $this->assertSame($expectedDate->format('Y-m-d'), $object->getDateProp()->format('Y-m-d'));
        $this->assertSame($defaultTimezone->getName(), $object->getDateProp()->getTimezone()->getName());
```

File: tests/Action/SetObjectFieldValueActionTest.php [hunk 6]
```diff
@@ -126,7 +126,25 @@ public function testSetObjectFieldValueAction(): void
...
+            [$default, $default],
+            [$custom->getName(), $custom],
+            [$custom, $custom],
+        ];
+    }
+
+    /**
+     * @dataProvider getTimeZones
+     */
+    public function testSetObjectFieldValueActionWithDate($timezone, \DateTimeZone $expected_timezone): void
```
Comment (phansys) [inline tests/Action/SetObjectFieldValueActionTest.php:147]:
```suggestion
    public function testSetObjectFieldValueActionWithDate($timezone, \DateTimeZone $expectedTimezone): void
```

File: tests/Action/SetObjectFieldValueActionTest.php [hunk 7]
```diff
@@ -171,9 +190,20 @@ public function testSetObjectFieldValueActionWithDate(): void
...
+        $this->assertInstanceOf(\DateTime::class, $object->getDateProp());
+        $this->assertSame($expected_date->format('Y-m-d'), $object->getDateProp()->format('Y-m-d'));
+        $this->assertSame($default_timezone->getName(), $object->getDateProp()->getTimezone()->getName());
     }
 
-    public function testSetObjectFieldValueActionWithDateTime(): void
+    /**
+     * @dataProvider getTimeZones
+     */
+    public function testSetObjectFieldValueActionWithDateTime($timezone, \DateTimeZone $expected_timezone): void
```
Comment (phansys) [inline tests/Action/SetObjectFieldValueActionTest.php:206]:
```suggestion
    public function testSetObjectFieldValueActionWithDateTime($timezone, \DateTimeZone $expectedTimezone): void
```

File: tests/Action/SetObjectFieldValueActionTest.php [hunk 8]
```diff
@@ -218,6 +249,14 @@ public function testSetObjectFieldValueActionWithDateTime(): void
...
 
         $this->assertSame(Response::HTTP_OK, $response->getStatusCode());
+
+        $default_timezone = new \DateTimeZone(date_default_timezone_get());
+        $expected_date = new \DateTime($request->query->get('value'), $expected_timezone);
+        $expected_date->setTimezone($default_timezone);
+
+        $this->assertInstanceOf(\DateTime::class, $object->getDatetimeProp());
+        $this->assertSame($expected_date->format('Y-m-d H:i:s'), $object->getDatetimeProp()->format('Y-m-d H:i:s'));
+        $this->assertSame($default_timezone->getName(), $object->getDatetimeProp()->getTimezone()->getName());
```
Comment (phansys) [inline tests/Action/SetObjectFieldValueActionTest.php:259]:
```suggestion
        $defaultTimezone = new \DateTimeZone(date_default_timezone_get());
        $expectedDate = new \DateTime($request->query->get('value'), $expectedTimezone);
        $expectedDate->setTimezone($defaultTimezone);

        $this->assertInstanceOf(\DateTime::class, $object->getDatetimeProp());
        $this->assertSame($expectedDate->format('Y-m-d H:i:s'), $object->getDatetimeProp()->format('Y-m-d H:i:s'));
        $this->assertSame($defaultTimezone->getName(), $object->getDatetimeProp()->getTimezone()->getName());
```

**Comment by peter-gribanov** [comment]
This PR broken. Renew PR #6167

**Pull request closed without merging**